lib/commands/interactive-cloud-command.ts

```diff
@@ -0,0 +1,53 @@
+import { isInteractive } from "../helpers";
+import { CloudOperationMessageTypes, CloudCommunicationEvents } from "../constants";
+
+export abstract class InteractiveCloudCommand implements ICommand {
+	public allowedParameters: ICommandParameter[];
+
+	constructor(private interactiveService: ICloudOperationService,
+		private $processService: IProcessService,
+		protected $errors: IErrors,
+		protected $logger: ILogger,
+		protected $prompter: IPrompter) {
+
+	}
+
+	public async execute(args: string[]): Promise<void> {
+		const messageHandler = async (msg: ICloudOperationMessage<any>) => {
+			if (msg.type === CloudOperationMessageTypes.CLOUD_OPERATION_INPUT_REQUEST) {
+				if (!isInteractive()) {
+					this.$errors.failWithoutHelp(`Input is required but the process is not interactive. "${msg.body.message}"`);
+				}
+
+				// Print one new line because we print the cloud operation output on the same line.
+				this.$logger.info();
+				let input = "";
+				if (msg.body.inputType === "password") {
+					input = await this.$prompter.getPassword(msg.body.message);
+				} else {
+					input = await this.$prompter.getString(msg.body.message);
+				}
+
+				try {
+					const inputBody: ICloudOperationInput = { inputType: msg.body.inputType, inputRequestId: msg.body.inputRequestId, content: input };
+					const cloudMessage: ICloudOperationMessage<ICloudOperationInput> = { type: CloudOperationMessageTypes.CLOUD_OPERATION_INPUT, cloudOperationId: msg.cloudOperationId, body: inputBody };
+					await this.interactiveService.sendCloudMessage(cloudMessage);
+				} catch (err) {
+					this.$logger.trace(`Can't send input to cloud operation ${msg.cloudOperationId}. Error is ${err}.`);
+				}
+			}
+		};
+
+		this.interactiveService.on(CloudCommunicationEvents.MESSAGE, messageHandler);
+
+		this.$processService.attachToProcessExitSignals(this, () => {
```

Comment:
Does this mean we will remove the handlers when the whole process exits? If so, then if hypothetically I call the same command twice, I will end up with two handlers attached

Reply:
Yes and no. The processService attaches only one listener to the process exit signals and it will execute the provided functions in it. So technically, we will have only one listener and we won't receive the memory leak message, but we will add more than one function which will be executed on process exit. This is not a problem because you need to call execute more than once to add more listeners and we want to remove the listener attached in the specific execute call, not all listeners.

Reply:
Hypothetically - yes. But since we are going to remove each handler, I prefer not to add additional logic to handle this hypothetical scenario.